src/tir/analysis/control_flow_graph.cc

```diff
@@ -0,0 +1,1642 @@
+            return visit_count_lookup[successor.index] == 0;
+          })) {
+        // Successors, if any, are unvisited.
+        return {};
+      } else if (block.successors.size() == 1) {
+        // Block has only a single successor
+        return states[0];
+      }
+
+      const auto& successor_a = block.successors[0];
+      const auto& successor_b = block.successors[1];
+
+      auto& post_a = states[0];
+      auto& post_b = states[1];
+
+      // During the first visit of a block, successor blocks may be
+      // unvisited, even though we preferentially visit later blocks
+      // first.  (e.g. During the first visit of the end of a For
+      // loop, the start of the For loop has not yet been visited.)
+      // If this is the case, assume the best-case scenario that all
+      // knowns are consistent, and rely on a later visit to
+      // resolve/remove any conflicts.
+      if (visit_count_lookup[successor_a.index] == 0) {
+        return post_b;
+      } else if (visit_count_lookup[successor_b.index] == 0) {
+        return post_a;
+      }
+
+      if (successor_a.post_condition && successor_b.post_condition) {
+        // The predicate can identify which predecessor block applies
+        // (e.g. i==0 for the first loop iteration, i>0 for remaining
+        // loop iterations).  Therefore, we can use all buffer
+        // constraints, conditional on having come from the
+        // predecessor that provides it.
+        post_a.AddCondition(successor_a.post_condition.value());
+        post_b.AddCondition(successor_b.post_condition.value());
+        post_a.Union(post_b, &analyzer);
+        return post_a;
+      } else {
+        // We don't know which predecessor applies.  Therefore, the
+        // only buffer constraints that can be used are those that
+        // appear in both successors.
+        post_a.Intersection(post_b, &analyzer);
+        return post_a;
+      }
+    }();
+
+    // Step 2: Collect knowns provided as a result of executing this block
+    auto unused_at_block_start = [&]() {
+      if (num_previous_visits >= max_revisits) {
+        return BufferState();
+      }
+      auto prior_state = block.unused_at_block_end;
+      prior_state.BackpropUnusedIndices(axis_var_lookup_, block.touch_points, &analyzer);
+      prior_state.RemoveFreeParameters(free_predicate_parameters_, &analyzer);
+      return prior_state;
+    }();
+
+    // Step 3: If any changes are made to the post knowns since the
+    // previous time we visited this block, mark the successor block
+    // as needing to be visited.
+    if (num_previous_visits == 0 ||
+        !unused_at_block_start.IsEquivalentTo(block.unused_at_block_start, &analyzer)) {
+      block.unused_at_block_start = std::move(unused_at_block_start);
+      for (const auto& pred : block.predecessors) {
+        to_visit.insert(pred.index);
+      }
+    }
+  }
+}
+
+bool ControlFlowGraph::IsOverwrittenWithoutEffect(const tir::BufferStore& store,
```

Comment:
It isn't right now, but will be used by https://github.com/apache/tvm/pull/13217 for identifying no-op stores.  It probably should have been in that PR instead of this one, but splitting things up into logically distinct PRs had been getting to be non-trivial.